packages/shell-api/src/bulk.ts

```diff
@@ -0,0 +1,217 @@
+      'arrayFilters method on fluent Bulk API is not currently supported.'
+    );
+  }
+
+  hint(hintDoc: Document): BulkFindOp {
+    assertArgsDefined(hintDoc);
+    this._hint = hintDoc;
+    return this;
+  }
+
+  remove(): Bulk {
+    this._parentBulk._batchCounts.nRemoveOps++;
+    this._innerFind.remove();
+    return this._parentBulk;
+  }
+
+  removeOne(): Bulk {
+    this._parentBulk._batchCounts.nRemoveOps++;
+    this._innerFind.removeOne();
+    return this._parentBulk;
+  }
+
+  replaceOne(replacement: Document): Bulk {
+    this._parentBulk._batchCounts.nUpdateOps++;
+    assertArgsDefined(replacement);
+    const op = { ...replacement };
+    if (this._hint) {
+      op.hint = this._hint;
+    }
+    this._innerFind.replaceOne(op);
+    return this._parentBulk;
+  }
+
+  updateOne(update: Document): Bulk {
+    this._parentBulk._batchCounts.nUpdateOps++;
+    assertArgsDefined(update);
+    const op = { ...update };
+    if (this._hint) {
+      op.hint = this._hint;
+    }
+    if (this._arrayFilters) {
+      op.arrayFilters = this._arrayFilters;
+    }
+    this._innerFind.updateOne(op);
+    return this._parentBulk;
+  }
+
+  update(update: Document): Bulk {
+    this._parentBulk._batchCounts.nUpdateOps++;
+    assertArgsDefined(update);
+    const op = { ...update };
+    if (this._hint) {
+      op.hint = this._hint;
+    }
+    if (this._arrayFilters) {
+      op.arrayFilters = this._arrayFilters;
+    }
+    this._innerFind.update(op);
+    return this._parentBulk;
+  }
+
+  upsert(): BulkFindOp {
+    assertArgsDefined();
+    this._innerFind.upsert();
+    return this;
+  }
+}
+
+
+@shellApiClassDefault
+@hasAsyncChild
+export default class Bulk extends ShellApiClass {
+  _mongo: Mongo;
+  _collection: any; // to avoid circular ref
+  _batchCounts: any;
+  _executed: boolean;
+  _batches: any;
+  _innerBulk: any;
+
+  constructor(collection, innerBulk) {
+    super();
+    this._collection = collection;
+    this._mongo = collection._mongo;
+    this._innerBulk = innerBulk;
+    this._batches = [];
+    this._batchCounts = {
+      nInsertOps: 0,
+      nUpdateOps: 0,
+      nRemoveOps: 0
+    };
+  }
+
+  /**
+   * Internal method to determine what is printed for this class.
+   */
+  _asPrintable(): string {
+    return this.tojson();
+  }
+
+  /**
+   * Internal helper for emitting collection API call events.
+   *
+   * @param methodName
+   * @param methodArguments
+   * @private
+   */
+  private _emitBulkApiCall(methodName: string, methodArguments: Document = {}): void {
+    this._mongo._internalState.emitApiCall({
+      method: methodName,
+      class: 'Bulk',
+      db: this._collection._database._name,
+      coll: this._collection._name,
+      arguments: methodArguments
+    });
+  }
+
+  @returnsPromise
+  async execute(writeConcern?: WriteConcern): Promise<BulkWriteResult> {
+    if (this._executed) {
+      throw new MongoshInvalidInputError('A bulk operation cannot be re-executed');
+    }
```

Comment:
to be precise in order to avoid race conditions in scripts `this._executed` should be set right here in the same "synchronous" context where is checked, or in any case before calling any async operation.

Reply:
We can get rid of this entirely because the driver will throw the same error. Or we could just set it to false in the constructor?

Reply:
I removed the error check but kept it it for modifying the batch count - it wouldn't make sense if calling execute twice changed the batch count. Let me know if you think there's a problem with that

Reply:
I think having `this._executed` there makes sense. my point was that we should not wait to set it to true though:

``` js
 async execute(writeConcern?: WriteConcern): Promise<BulkWriteResult> {
    if (this._executed) {
      throw new MongoshInvalidInputError('A bulk operation cannot be re-executed');
    }
    this._executed = true;
   ...
```

which is different from:

``` js
    if (this._executed) {
      throw new MongoshInvalidInputError('A bulk operation cannot be re-executed');
    }

    await doSomething();  // <--

    this._executed = true;
```

Reply:
But shouldn't `this._executed` only be set true if the driver successfully executes the bulk op? That's why it's after the `await`